### PR TITLE
Add HVAC Ray example

### PR DESCRIPTION
**Pull request by vinayhanumaiah**
*Issue #, if available:* N/A

*Description of changes:*
This PR adds an example of using RL for HVAC control using EnergyPlus. Conceptually, it's similar to existing [example](https://github.com/aws/amazon-sagemaker-examples/tree/master/reinforcement_learning/rl_hvac_coach_energyplus), however, this example uses Ray RLlib (SOTA RL framework), APEX-DDPG (one of the SOTA RL algorithm), and also uses distributed rollout workers to fully utilize all CPU cores.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: reinforcement_learning/rl_hvac_ray_energyplus/source/rollout.py [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self._shelf:
+            # Close the shelf file, and store the number of episodes for ease
+            self._shelf["num_episodes"] = self._num_episodes
+            self._shelf.close()
+        elif self._outfile and not self._use_shelve:
+            # Dump everything as one big pickle:
+            pickle.dump(self._rollouts, open(self._outfile, "wb"))
```
[the remark this answers is not in the document]
Comment (vinayhanumaiah) [reply reinforcement_learning/rl_hvac_ray_energyplus/source/rollout.py:136]: Added try-finally block.

File: reinforcement_learning/rl_hvac_ray_energyplus/source/hvac_ray_launcher.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            from ray.rllib.agents.agent import get_agent_class
+        cls = get_agent_class(algorithm)
+        with open(os.path.join(MODEL_OUTPUT_DIR, "params.json")) as config_json:
+            config = json.load(config_json)
+        use_torch = config.get("use_pytorch", False)
+        if not use_torch:
+            if 'callbacks' in config:
+                callback_cls_str = config['callbacks']
+                callback_cls = callback_cls_str.split("'")[-2].split(".")[-1]
+                config['callbacks'] = eval(callback_cls)
```
[the remark this answers is not in the document]
Comment (vinayhanumaiah) [reply reinforcement_learning/rl_hvac_ray_energyplus/source/hvac_ray_launcher.py]: Replaced eval with ast.literal_eval().

File: reinforcement_learning/rl_hvac_ray_energyplus/source/rollout.py [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+        done = False
+        reward_total = 0.0
+        episode_steps = 0
+        while not done and keep_going(steps, num_steps, episodes,
+                                      num_episodes):
+            multi_obs = obs if multiagent else {_DUMMY_AGENT_ID: obs}
+            action_dict = {}
+            for agent_id, a_obs in multi_obs.items():
+                if a_obs is not None:
+                    policy_id = mapping_cache.setdefault(
```
[the remark this answers is not in the document]
Comment (vinayhanumaiah) [reply reinforcement_learning/rl_hvac_ray_energyplus/source/rollout.py:418]: In this case, we cannot pre-configure default values and it happens in run-time. Leaving it as it is.

**Comment by vinayhanumaiah** [comment]
From the logs, it looks like the build is failing while running docker build, But I can run with no issues on SageMaker. Can you let me know what's the exact issue?

**Review by ajaykarpur — APPROVED** [review]
Manually overriding because of Docker issues in CI.

**Pull request merged**